show-callsites: add an analysis to dump a CFG

nmanthey opened the pull request:
This more or less dumps a CFG, and adds a line that shows all functions
whose address has been consumed. The output is in CSV, so that it can
be consumed easily by further automation.

tautschnig commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,100 @@
+#include <goto-programs/compute_called_functions.h>
+#include <goto-programs/goto_model.h>
+#include <langapi/language.h>
+#include <langapi/mode.h>
```
Do you genuinely need those langapi headers here?

tautschnig commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,100 @@
+#include <util/message.h>
+#include <util/prefix.h>
+
+#include <algorithm>
```
Seems unused

tautschnig commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,100 @@
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
```
Seems unused

tautschnig commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,100 @@
+#include <algorithm>
+#include <fstream>
+#include <sstream>
+#include <string>
```
Already included by the header file

tautschnig commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,100 @@
+    const goto_functionst::function_mapt::const_iterator func_it =
+      goto_model.goto_functions.function_map.find(f_it->first);
+
+    if(func_it == goto_model.goto_functions.function_map.end())
```
Isn't `func_it` the same as `f_it`?

nmanthey replied:
I will drop the whole block.

tautschnig commented on src/goto-analyzer/show_callsites.h:
```diff
@@ -0,0 +1,23 @@
+#ifndef CPROVER_GOTO_ANALYZER_SHOW_CALLSITES_H
+#define CPROVER_GOTO_ANALYZER_SHOW_CALLSITES_H
+
+#include "goto-programs/goto_model.h"
```
A forward declaration of `class goto_modelt;` would suffice here.

nmanthey replied:
will fix.

tautschnig reviewed (commented):
A test case would be nice as it would also serve as documentation of that the result will look like.

kroening commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,100 @@
+void show_callsites(
+  const goto_modelt &goto_model,
+  message_handlert &ui,
+  const std::string &output_filename)
```
The message handler isn't used?
We would usually call it "message_handler".

nmanthey replied:
i'll drop it, that's left over from copy&paste

kroening commented on src/goto-analyzer/show_callsites.h:
```diff
@@ -0,0 +1,23 @@
+/*******************************************************************\
+
+Module: Show symbols that participate in shadowing
+
```
The module description is somewhat mysterious.

nmanthey replied:
I agree that this description does not match the actual code. That's a copy&paste left over from another analysis prototype. I'll fix it.

kroening commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -811,6 +818,8 @@ void goto_analyzer_parse_optionst::help()
     " --reachable-functions        list functions reachable from the entry point\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --show-callsites=<filename>  generate file that lists functions call sites (includes functions for pointers)\n"
     "\n"
```
We usually rely on a separate option to set the output format and file name; we already have:

Output options:
 --text file_name             output results in plain text to given file
 --json file_name             output results in JSON format to given file
 --xml file_name              output results in XML format to given file
 --dot file_name              output results in DOT format to given file

You could add --csv file_name

nmanthey replied:
I disagree here, and will not add a new --csv option. Given the surroundings in goto-analyzer control parts of the output of the tool, writing to a file is a stable option. This way, I get only the output I need for further processing.

nmanthey commented:
I'll fix the redundant headers, and add a test next.

nmanthey commented:
I adapted the code according to the above comments, and added a test case. Note that this test case is sensitive to changes in function pointer removal! I furthermore allowed to re-route the output to stdout, in case '-' is given as argument for the output file name.

tautschnig commented on regression/goto-analyzer/show-callsites/test.desc:
```diff
@@ -0,0 +1,18 @@
+--show-callsites -
+^EXIT=0$
+^SIGNAL=0$
+__BY_ADDRESS__,f2,f4,f1,f5,f0
```
How about: `^__BY_ADDRESS__,(f[01245],){4}(f[01245])$`? Then do similarly for the below.

nmanthey commented:
I adapted the pattern in test.desc. This is more stable now, but still misses the case when "func" will be on a different position (which now is less likely, but might still bite in the future).

peterschrammel commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,96 @@
+/** show_callsites writes a call graph to CSV
+ *
+ * This function creates a map between functions and consumed functions, and
+ * stores this relationship in a CSV file, to be processed outside of
```
Do we now have a 4th output format? Why is Json or XML not good enough?

martin-cs replied:
I hate to say it but... yeah... we should either fully commit to having CSV as a fourth format (which I don't think is a great idea) or use one of the existing ones.  We don't need to make auxiliary tooling any more confused than it is now.

nmanthey commented:
For CSV, I can use cli tools to easily extract information from the output,
because it is line based. That is hard for json and xml, depending on who
prints the output its at least "ugly".

I can use another json to csv conversion, if you insist on using json or
xml.

tautschnig commented:
@nmanthey Would `jq` possibly do the trick? Yes, extra processing work is painful, but yet-another-custom-format indeed seems undesirable.

nmanthey commented:
The code now prints json. Getting back to CSV is basically dropping all {}", and replace ':' with ',' as well.

tautschnig commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,115 @@
+  std::stringstream function_line;
+
+  // handle functions whose address is used
+  *output_stream << "{" << std::endl;
```
Is this constructing JSON-formatting string output? I'd highly recommend using the classes from `util/json.h` instead and then just do `*output_stream << your_json_object`.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 6eec362ec20b6151fad00ae301ecf4a8090da596).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77435621
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj commented:
Hi Norbert. The reason for the failure that is mentioned in my comment is that our CI cannot fetch forks for devs that are not in the list of contributors. I checked your PR manually though and it's all fine.

polgreen commented on regression/goto-analyzer/show-callsites/main.c:
```diff
@@ -0,0 +1,61 @@
+}
+void f4(void)
+{
+  f5();
```
This doesn't compile, because `f5` is not declared yet.  (It does compile with goto-cc, maybe a bug in goto-cc, or expected behaviour?)

polgreen commented on regression/goto-analyzer/show-callsites/main.c:
```diff
@@ -0,0 +1,61 @@
+void func()
+{
+  fp_tbl[0]();
+  fp_tbl[1]();
```
`fp_tbl[1]` is a null pointer; is this deliberate?

nmanthey replied:
As that program is only a test case, the segfault is fine. I'll stil fix the compilation error.

nmanthey commented:
The latest version uses the CBMC classes to generate json.

peterschrammel commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -681,6 +682,11 @@ int goto_analyzer_parse_optionst::perform_analysis(const optionst &options)
+  if(cmdline.isset("show-callsites"))
+  {
+    show_callsites(goto_model, cmdline.get_value("show-callsites"));
+    return 0;
```
use an appropriate code from `util/exit_codes.h`

nmanthey replied:
will fix

peterschrammel commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+    by_address.push_back(json_stringt(str_name));
+  }
+  if(by_address.size())
+    result["__BY_ADDRESS__"] = by_address;
```
keys should be camelCase, eg. `"byAddress"`

nmanthey replied:
this field is a special case that hopefully does not occur as a real function name. Therefore, I'll keep it like this.

peterschrammel commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+    result["__BY_ADDRESS__"] = by_address;
+
+  // go through call graph
+  int functions = 0;
```
seems unused

nmanthey replied:
will drop

peterschrammel commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+    if(
+      has_prefix(str_name, CPROVER_PREFIX) ||
+      (str_name.find('$') != std::string::npos))
+      continue;
```
```
{
  continue;
}
```
would be more readable after multi-line conditionals

nmanthey replied:
This code has been reformatted by clang-formater. I can test what happens if I add these brackets.

peterschrammel commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    functions++;
+    result[id2string(f_it->first)] = function_line;
```
Keys should be camelCase. Hm... I see now... That's a bit atypical for a JSON file. Typically it should be rather structured as follows:
```
[
  {
     "name" : "my_caller",
     "callees" : [ "my_callee1", "my_callee2" ]
  },
  ...
]
```

nmanthey replied:
I would like to stick to the "caller" -> callees structure, as that represents the graph that this structure represents. I already moved away from CSV, i want to keep at least "have each adjacency list in a single line, starting with the caller".

LAJW replied:
I agree with @nmanthey. Let's store map as a map.

LAJW commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  std::ostream *output_stream =
+    output_filename == "-" ? &std::cout : &output_file;
```
I'm pretty sure that the message handler should be used here instead of `std::cout`.

LAJW replied:
Also, since you're using ternary operator (👍 for that), you can also store `output_stream` as a reference rather than as a pointer.

LAJW commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+    const std::string str_name = id2string(name);
+    if(
+      has_prefix(str_name, CPROVER_PREFIX) ||
+      (str_name.find('$') != std::string::npos))
```
Unnecessary braces, I'd think that everyone knows order of operators. Also, since you're checking this twice, you could put it into a static function. That would reduce the size of the if statement, so then you wouldn't be forced to wrap `continue` in braces, which in the end means - less lines!

LAJW commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+  json_arrayt by_address;
+  for(const auto &name : excluded_functions)
+  {
+    const std::string str_name = id2string(name);
```
FYI you can inline it. `id2string` returns a reference to a string (irep_idt actually does store std::string), so you don't have to declare that variable.

LAJW commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+
+    json_arrayt function_line;
+    const goto_programt &program = f_it->second.body;
+    forall_goto_program_instructions(i_it, program)
```
I'm pretty sure, you can replace that macro with a ranged-for.

LAJW commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+        continue;
+
+      const auto &function = to_code_function_call(i_it->code).function();
+      auto &se = to_symbol_expr(function).get_identifier();
```
You might've forgotten const. Otherwise, 👍

LAJW commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,101 @@
+  {
+    output_file.open(output_filename);
+    if(!output_file.is_open())
+      throw "failed to open callsite output file";
```
Please, throw `std::runtime_error` or something similar, rather than a string. We often throw strings to signal an assertion (which is an antipattern, we have `INVARIANT` for that), but in this case, it'd make more sense to display a pretty message to the user, rather than a scary stack trace.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 72197f7ca7608e1cce7556b930ed2300bb1bb0d2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78286528
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: e24e0d708ff5e5cf10d974eab7cf345b10918359).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78297533
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

nmanthey commented:
I tried the suggested replacements, but they all failed. For the output type, a pointer has to be used to allow chosing from different stream types. The message handler does not always come with a stream, and therefore cannot be used there. Finally, the for loops cannot be "just" replaced, as the underlying data structures would require some additional work as well. My latest version addresses the other comments.

LAJW commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  std::ostream *output_stream =
+    output_filename == "-" ? &std::cout : &output_file;
```
If you can't use the message handler, then store this variable as a reference rather than a pointer.

nmanthey replied:
When using a reference, the different types cannot be assigned by the ternary operator, as inheritance requires pointers.

LAJW replied:
References are pointers. Inheritance works with references. The only difference is that references have to be initialized. Just try it, you'll see :-)

LAJW reviewed (approved):
Great work! Thanks for trying. We were trying to get rid of those macros for quite a while now. More often than not they are prolonged because of the force of habit. But since this is not the case here, then that's fine. We'll kill them later.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 8ba7caaa475ae542813813680c5e531e21b03faf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78610673
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

peterschrammel commented:
@nmanthey, There's a failing Windows test: 
```
GOTO-ANALYSER version 5.9 (dc82c0042-dirty) 64-bit x86_64 windows
Parsing main.c
main.c
Converting
Type-checking main
Generating GOTO Program
Removing function pointers and virtual functions
file main.c line 58 function main: replacing function pointer by 5 possible targets
file main.c line 46 function func: replacing function pointer by 0 possible targets
Partial Inlining
{
  "__BY_ADDRESS__": [ "f0", "f4", "f2", "f5", "f1" ],
  "__acrt_iob_func": [ ],
  "__local_stdio_printf_options": [ ],
  "__local_stdio_scanf_options": [ ],
  "__stdio_common_vfprintf": [ ],
  "__stdio_common_vfprintf_p": [ ],
  "__stdio_common_vfprintf_s": [ ],
  "__stdio_common_vfscanf": [ ],
  "__stdio_common_vfwprintf": [ ],
  "__stdio_common_vfwprintf_p": [ ],
  "__stdio_common_vfwprintf_s": [ ],
  "__stdio_common_vfwscanf": [ ],
  "__stdio_common_vsnprintf_s": [ ],
  "__stdio_common_vsnwprintf_s": [ ],
  "__stdio_common_vsprintf": [ ],
  "__stdio_common_vsprintf_p": [ ],
  "__stdio_common_vsprintf_s": [ ],
  "__stdio_common_vsscanf": [ ],
  "__stdio_common_vswprintf": [ ],
  "__stdio_common_vswprintf_p": [ ],
  "__stdio_common_vswprintf_s": [ ],
  "__stdio_common_vswscanf": [ ],
  "__swprintf_l": [ "__va_start", "__stdio_common_vswprintf" ],
  "__va_start": [ ],
  "__vswprintf_l": [ "__stdio_common_vswprintf" ],
  "_fprintf_l": [ "__va_start", "__stdio_common_vfprintf" ],
  "_fprintf_p": [ "__va_start", "__stdio_common_vfprintf_p" ],
  "_fprintf_p_l": [ "__va_start", "__stdio_common_vfprintf_p" ],
  "_fprintf_s_l": [ "__va_start", "__stdio_common_vfprintf_s" ],
  "_fscanf_l": [ "__va_start", "__stdio_common_vfscanf" ],
  "_fscanf_s_l": [ "__va_start", "__stdio_common_vfscanf" ],
  "_fwprintf_l": [ "__va_start", "__stdio_common_vfwprintf" ],
  "_fwprintf_p": [ "__va_start", "__stdio_common_vfwprintf_p" ],
  "_fwprintf_p_l": [ "__va_start", "__stdio_common_vfwprintf_p" ],
  "_fwprintf_s_l": [ "__va_start", "__stdio_common_vfwprintf_s" ],
  "_fwscanf_l": [ "__va_start", "__stdio_common_vfwscanf" ],
  "_fwscanf_s_l": [ "__va_start", "__stdio_common_vfwscanf" ],
  "_printf_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfprintf" ],
  "_printf_p": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfprintf_p" ],
  "_printf_p_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfprintf_p" ],
  "_printf_s_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfprintf_s" ],
  "_scanf_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfscanf" ],
  "_scanf_s_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfscanf" ],
  "_scprintf": [ "__va_start", "__stdio_common_vsprintf" ],
  "_scprintf_l": [ "__va_start", "__stdio_common_vsprintf" ],
  "_scprintf_p": [ "__va_start", "__stdio_common_vsprintf_p" ],
  "_scprintf_p_l": [ "__va_start", "__stdio_common_vsprintf_p" ],
  "_scwprintf": [ "__va_start", "__stdio_common_vswprintf" ],
  "_scwprintf_l": [ "__va_start", "__stdio_common_vswprintf" ],
  "_scwprintf_p": [ "__va_start", "__stdio_common_vswprintf_p" ],
  "_scwprintf_p_l": [ "__va_start", "__stdio_common_vswprintf_p" ],
  "_snprintf": [ "__va_start", "__stdio_common_vsprintf" ],
  "_snprintf_c": [ "__va_start", "__stdio_common_vsprintf" ],
  "_snprintf_c_l": [ "__va_start", "__stdio_common_vsprintf" ],
  "_snprintf_l": [ "__va_start", "__stdio_common_vsprintf" ],
  "_snprintf_s": [ "__va_start", "__stdio_common_vsnprintf_s" ],
  "_snprintf_s_l": [ "__va_start", "__stdio_common_vsnprintf_s" ],
  "_snscanf": [ "__va_start", "__stdio_common_vsscanf" ],
  "_snscanf_l": [ "__va_start", "__stdio_common_vsscanf" ],
  "_snscanf_s": [ "__va_start", "__stdio_common_vsscanf" ],
  "_snscanf_s_l": [ "__va_start", "__stdio_common_vsscanf" ],
  "_snwprintf": [ "__va_start", "__stdio_common_vswprintf" ],
  "_snwprintf_l": [ "__va_start", "__stdio_common_vswprintf" ],
  "_snwprintf_s": [ "__va_start", "__stdio_common_vsnwprintf_s" ],
  "_snwprintf_s_l": [ "__va_start", "__stdio_common_vsnwprintf_s" ],
  "_snwscanf": [ "__va_start", "__stdio_common_vswscanf" ],
  "_snwscanf_l": [ "__va_start", "__stdio_common_vswscanf" ],
  "_snwscanf_s": [ "__va_start", "__stdio_common_vswscanf" ],
  "_snwscanf_s_l": [ "__va_start", "__stdio_common_vswscanf" ],
  "_sprintf_l": [ "__va_start", "__stdio_common_vsprintf" ],
  "_sprintf_p": [ "__va_start", "__stdio_common_vsprintf_p" ],
  "_sprintf_p_l": [ "__va_start", "__stdio_common_vsprintf_p" ],
  "_sprintf_s_l": [ "__va_start", "__stdio_common_vsprintf_s" ],
  "_sscanf_l": [ "__va_start", "__stdio_common_vsscanf" ],
  "_sscanf_s_l": [ "__va_start", "__stdio_common_vsscanf" ],
  "_swprintf": [ "__va_start", "__stdio_common_vswprintf" ],
  "_swprintf_c": [ "__va_start", "__stdio_common_vswprintf" ],
  "_swprintf_c_l": [ "__va_start", "__stdio_common_vswprintf" ],
  "_swprintf_l": [ "__va_start", "__stdio_common_vswprintf" ],
  "_swprintf_p": [ "__va_start", "__stdio_common_vswprintf_p" ],
  "_swprintf_p_l": [ "__va_start", "__stdio_common_vswprintf_p" ],
  "_swprintf_s_l": [ "__va_start", "__stdio_common_vswprintf_s" ],
  "_swscanf_l": [ "__va_start", "__stdio_common_vswscanf" ],
  "_swscanf_s_l": [ "__va_start", "__stdio_common_vswscanf" ],
  "_vfprintf_l": [ "__stdio_common_vfprintf" ],
  "_vfprintf_p": [ "__stdio_common_vfprintf_p" ],
  "_vfprintf_p_l": [ "__stdio_common_vfprintf_p" ],
  "_vfprintf_s_l": [ "__stdio_common_vfprintf_s" ],
  "_vfscanf_l": [ "__stdio_common_vfscanf" ],
  "_vfscanf_s_l": [ "__stdio_common_vfscanf" ],
  "_vfwprintf_l": [ "__stdio_common_vfwprintf" ],
  "_vfwprintf_p": [ "__stdio_common_vfwprintf_p" ],
  "_vfwprintf_p_l": [ "__stdio_common_vfwprintf_p" ],
  "_vfwprintf_s_l": [ "__stdio_common_vfwprintf_s" ],
  "_vfwscanf_l": [ "__stdio_common_vfwscanf" ],
  "_vfwscanf_s_l": [ "__stdio_common_vfwscanf" ],
  "_vprintf_l": [ "__acrt_iob_func", "__stdio_common_vfprintf" ],
  "_vprintf_p": [ "__acrt_iob_func", "__stdio_common_vfprintf_p" ],
  "_vprintf_p_l": [ "__acrt_iob_func", "__stdio_common_vfprintf_p" ],
  "_vprintf_s_l": [ "__acrt_iob_func", "__stdio_common_vfprintf_s" ],
  "_vscanf_l": [ "__acrt_iob_func", "__stdio_common_vfscanf" ],
  "_vscanf_s_l": [ "__acrt_iob_func", "__stdio_common_vfscanf" ],
  "_vscprintf": [ "__stdio_common_vsprintf" ],
  "_vscprintf_l": [ "__stdio_common_vsprintf" ],
  "_vscprintf_p": [ "__stdio_common_vsprintf_p" ],
  "_vscprintf_p_l": [ "__stdio_common_vsprintf_p" ],
  "_vscwprintf": [ "__stdio_common_vswprintf" ],
  "_vscwprintf_l": [ "__stdio_common_vswprintf" ],
  "_vscwprintf_p": [ "__stdio_common_vswprintf_p" ],
  "_vscwprintf_p_l": [ "__stdio_common_vswprintf_p" ],
  "_vsnprintf": [ "__stdio_common_vsprintf" ],
  "_vsnprintf_c": [ "__stdio_common_vsprintf" ],
  "_vsnprintf_c_l": [ "__stdio_common_vsprintf" ],
  "_vsnprintf_l": [ "__stdio_common_vsprintf" ],
  "_vsnprintf_s": [ "__stdio_common_vsnprintf_s" ],
  "_vsnprintf_s_l": [ "__stdio_common_vsnprintf_s" ],
  "_vsnwprintf": [ "__stdio_common_vswprintf" ],
  "_vsnwprintf_l": [ "__stdio_common_vswprintf" ],
  "_vsnwprintf_s": [ "__stdio_common_vsnwprintf_s" ],
  "_vsnwprintf_s_l": [ "__stdio_common_vsnwprintf_s" ],
  "_vsnwscanf_l": [ "__stdio_common_vswscanf" ],
  "_vsnwscanf_s_l": [ "__stdio_common_vswscanf" ],
  "_vsprintf_l": [ "__stdio_common_vsprintf" ],
  "_vsprintf_p": [ "__stdio_common_vsprintf_p" ],
  "_vsprintf_p_l": [ "__stdio_common_vsprintf_p" ],
  "_vsprintf_s_l": [ "__stdio_common_vsprintf_s" ],
  "_vsscanf_l": [ "__stdio_common_vsscanf" ],
  "_vsscanf_s_l": [ "__stdio_common_vsscanf" ],
  "_vswprintf": [ "__stdio_common_vswprintf" ],
  "_vswprintf_c": [ "__stdio_common_vswprintf" ],
  "_vswprintf_c_l": [ "__stdio_common_vswprintf" ],
  "_vswprintf_l": [ "__stdio_common_vswprintf" ],
  "_vswprintf_p": [ "__stdio_common_vswprintf_p" ],
  "_vswprintf_p_l": [ "__stdio_common_vswprintf_p" ],
  "_vswprintf_s_l": [ "__stdio_common_vswprintf_s" ],
  "_vswscanf_l": [ "__stdio_common_vswscanf" ],
  "_vswscanf_s_l": [ "__stdio_common_vswscanf" ],
  "_vwprintf_l": [ "__acrt_iob_func", "__stdio_common_vfwprintf" ],
  "_vwprintf_p": [ "__acrt_iob_func", "__stdio_common_vfwprintf_p" ],
  "_vwprintf_p_l": [ "__acrt_iob_func", "__stdio_common_vfwprintf_p" ],
  "_vwprintf_s_l": [ "__acrt_iob_func", "__stdio_common_vfwprintf_s" ],
  "_vwscanf_l": [ "__acrt_iob_func", "__stdio_common_vfwscanf" ],
  "_vwscanf_s_l": [ "__acrt_iob_func", "__stdio_common_vfwscanf" ],
  "_wprintf_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwprintf" ],
  "_wprintf_p": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwprintf_p" ],
  "_wprintf_p_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwprintf_p" ],
  "_wprintf_s_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwprintf_s" ],
  "_wscanf_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwscanf" ],
  "_wscanf_s_l": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwscanf" ],
  "f0": [ ],
  "f1": [ ],
  "f2": [ ],
  "f3": [ ],
  "f4": [ ],
  "f5": [ "f4" ],
  "f6": [ ],
  "fprintf": [ "__va_start", "__stdio_common_vfprintf" ],
  "fprintf_s": [ "__va_start", "__stdio_common_vfprintf_s" ],
  "fscanf": [ "__va_start", "__stdio_common_vfscanf" ],
  "fscanf_s": [ "__va_start", "__stdio_common_vfscanf" ],
  "func": [ "f2" ],
  "fwprintf": [ "__va_start", "__stdio_common_vfwprintf" ],
  "fwprintf_s": [ "__va_start", "__stdio_common_vfwprintf_s" ],
  "fwscanf": [ "__va_start", "__stdio_common_vfwscanf" ],
  "fwscanf_s": [ "__va_start", "__stdio_common_vfwscanf" ],
  "main": [ "func", "f4", "f0", "f1", "f5", "f2" ],
  "printf": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfprintf" ],
  "printf_s": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfprintf_s" ],
  "scanf": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfscanf" ],
  "scanf_s": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfscanf" ],
  "snprintf": [ "__va_start", "__stdio_common_vsprintf" ],
  "sprintf": [ "__va_start", "__stdio_common_vsprintf" ],
  "sprintf_s": [ "__va_start", "__stdio_common_vsprintf_s" ],
  "sscanf": [ "__va_start", "__stdio_common_vsscanf" ],
  "sscanf_s": [ "__va_start", "__stdio_common_vsscanf" ],
  "swprintf": [ "__va_start", "__stdio_common_vswprintf" ],
  "swprintf_s": [ "__va_start", "__stdio_common_vswprintf_s" ],
  "swscanf": [ "__va_start", "__stdio_common_vswscanf" ],
  "swscanf_s": [ "__va_start", "__stdio_common_vswscanf" ],
  "vfprintf": [ "__stdio_common_vfprintf" ],
  "vfprintf_s": [ "__stdio_common_vfprintf_s" ],
  "vfscanf": [ "__stdio_common_vfscanf" ],
  "vfscanf_s": [ "__stdio_common_vfscanf" ],
  "vfwprintf": [ "__stdio_common_vfwprintf" ],
  "vfwprintf_s": [ "__stdio_common_vfwprintf_s" ],
  "vfwscanf": [ "__stdio_common_vfwscanf" ],
  "vfwscanf_s": [ "__stdio_common_vfwscanf" ],
  "vprintf": [ "__acrt_iob_func", "__stdio_common_vfprintf" ],
  "vprintf_s": [ "__acrt_iob_func", "__stdio_common_vfprintf_s" ],
  "vscanf": [ "__acrt_iob_func", "__stdio_common_vfscanf" ],
  "vscanf_s": [ "__acrt_iob_func", "__stdio_common_vfscanf" ],
  "vsnprintf": [ "__stdio_common_vsprintf" ],
  "vsnprintf_s": [ "__stdio_common_vsnprintf_s" ],
  "vsprintf": [ "__stdio_common_vsprintf" ],
  "vsprintf_s": [ "__stdio_common_vsprintf_s" ],
  "vsscanf": [ "__stdio_common_vsscanf" ],
  "vsscanf_s": [ "__stdio_common_vsscanf" ],
  "vswprintf": [ "__stdio_common_vswprintf" ],
  "vswprintf_s": [ "__stdio_common_vswprintf_s" ],
  "vswscanf": [ "__stdio_common_vswscanf" ],
  "vswscanf_s": [ "__stdio_common_vswscanf" ],
  "vwprintf": [ "__acrt_iob_func", "__stdio_common_vfwprintf" ],
  "vwprintf_s": [ "__acrt_iob_func", "__stdio_common_vfwprintf_s" ],
  "vwscanf": [ "__acrt_iob_func", "__stdio_common_vfwscanf" ],
  "vwscanf_s": [ "__acrt_iob_func", "__stdio_common_vfwscanf" ],
  "wprintf": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwprintf" ],
  "wprintf_s": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwprintf_s" ],
  "wscanf": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwscanf" ],
  "wscanf_s": [ "__va_start", "__acrt_iob_func", "__stdio_common_vfwscanf" ]
}
EXIT=0
SIGNAL=0


Failed test.desc lines:
^  "printf": \[ \] [FAILED]
make[2]: *** [Makefile:4: test] Error 1
make[2]: Leaving directory '/cygdrive/c/codebuild/tmp/src068426683/src/github.com/diffblue/cbmc/regression/goto-analyzer'
make[1]: *** [Makefile:41: goto-analyzer] Error 1
make[1]: Leaving directory '/cygdrive/c/codebuild/tmp/src068426683/src/github.com/diffblue/cbmc/regression'
make: *** [Makefile:33: test] Error 1
make: Leaving directory '/cygdrive/c/codebuild/tmp/src068426683/src/github.com/diffblue/cbmc/regression'
```

peterschrammel reviewed (approved):
LGTM modulo CI passing

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: dc82c0042314e830c9deb99b0ba086c7a3712498).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80942250
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

martin-cs commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -808,6 +814,8 @@ void goto_analyzer_parse_optionst::help()
     // NOLINTNEXTLINE(whitespace/line_length)
     " --reachable-functions        list functions reachable from the entry point\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --show-callsites filename    list function call graph as CSV (includes functions for pointers, '-' for stdout)\n"
```
Sorry but I don't think this should go here.  The idea behind goto-analyzer is to run an abstract interpreter and then perform some kind of task based on the domains.  So --reachable-functions is relative to some analysis.  Show call sites is purely syntactic, so it shouldn't be listed with these.  IMHO it should either be:

(Preferable) A. In goto-instrument's diagnosis options, along with --call-graph and --reachable-call-graph (speaking of which, how is this different?)

B. goto-analyze's "Specific analyses" which are things that conceptually kind of fit in goto-analyzer but don't allow the parameterisation of domain, abstract interpreter, etc.

nmanthey replied:
It's different in (1) it does not print CBMC functions that do not occur in the actual source, and (2) it prints the list of all functions whose address is consumed, so that I now that I have to treat those functions specially when running follow up analysis.

I do not understand why that should be moved into goto-instrument. The code displays something about the goto binary. I do not want to implement the follow up analysis in CPROVER, as that can be done Python. I just want to make relevant parts of the goto-program consumable for Python. If that should be moved to goto-instrument, I can do it.

martin-cs commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,98 @@
+/*******************************************************************\
+
+Module: Write a function call graph to a CSV file
```
Not sure it is generating CSV any more...

nmanthey replied:
True, left over, will fix it all over the place. I'll use JSON.

martin-cs commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,98 @@
+         name.find('$') != std::string::npos;
+}
+
+/** show_callsites writes a call graph to CSV
```
Again CSV

martin-cs commented on src/goto-analyzer/show_callsites.cpp:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  std::ostream &output_stream =
+    output_filename == "-" ? std::cout : output_file;
```
I tend to feel this should be done in the *_parse_options code.

nmanthey replied:
Will move it.

martin-cs reviewed (changes requested):
Sorry to come to this late and request major changes but I think I was on leave when this was first raised.

martin-cs commented:
Would it be possible to add this functionality as the JSON output of ```goto-instrument --call-graph```:

```
$ ./goto-instrument/goto-instrument --call-graph call-graph-test.goto
Reading GOTO program from `call-graph-test.goto'
Function Pointer Removal
file call-graph-test.c line 49 function main: replacing function pointer by 5 possible targets
file call-graph-test.c line 41 function func: replacing function pointer by 0 possible targets
Virtual function removal
Cleaning inline assembler statements
main -> func
main -> f4
main -> f2
main -> f1
main -> f5
main -> f0
__CPROVER__start -> __CPROVER_initialize
__CPROVER__start -> main
f5 -> f4
func -> f2
```

nmanthey commented:
I could extend the call graph to print JSON, but this change is more about adding more specific output which is currently not provided in the call graph class. I will not touch the existing functionality.

I'll rewrite the new file to (1) use the call graph class including the output, and will manually introduce the artificial node "__BY_ADDRESS" and all relevant edges, as well as drop all CPROVER functions that are present in the initial call graph. Finally, I'll use the existing output method.

I'll adapt my Python script that consumes the output accordingly, again.

nmanthey committed:
call_graph: add address taken functions

Depending on the analysis on the call graph, it might be relevant
to treat functions special if their address has been taken. To
cover this case, add a method that will add edges from a new
artificial function to all functions whose address is taken. The
name of the new node can be specified via a parameter. The functions
to add are generated by using the compute_address_taken function.

nmanthey committed:
show_callsites: dump CFG for further processing

The current methods to dump a CFG contain internal functions, and
do not print functions whose address is taken, as well as print
only to std::cout. To process the CFG in a stable way, allow to
print it to a file, and drop artificial functions, as well as add
an artificial node "__BY_ADDRESS__" which calls all functions whose
address has been consumed.

nmanthey commented:
I just noticed another difference: the call graph as it's currently implemented will not mention the function "printk" at all. However, I do want to see printk in the output, so going the call graph route is also not easy.

nmanthey committed:
tests: add show_callsites test

This test checks whether the expected call graph is printed, as well
as the functions whose address is consumed. Futhermore, we want to
make sure that no internal function names appear in the graph.

nmanthey committed:
call_graph: drop internal functions

When constructing the call graph, functions of CPROVER are present
as well. However, when the call graph should be consumed, these
functions are not required. Hence, we add a method that can drop
these functions from the nodes and edges of the graph again.

nmanthey commented:
I moved the functionality to goto-instrument, and extended the call graph to (1) drop internal functions and (2) add extra edges for functions whose address is consumed. Piping the output to a file is still there.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 901fdf48308e27fee598f3a2cd9facb502d4c67d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85423329
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

nmanthey commented:
@martin-cs I reworked the code and you seem to be the only reviewer blocking the merge. Is there anything else that should be done? Otherwise, I'd be happy if somebody would merge the changes so that this PR can be closed. Thanks!

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: e506baf40e50e5eabfe94a5d38d5aaf42d8fd126).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86620921
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

TGWDB commented:
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.
